main.go

```diff
@@ -0,0 +1,258 @@
+		procUptime: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "proc_uptime"),
+			"Number of seconds since processor started.",
+			[]string{"name", "pid"},
+			nil,
+		),
+		procMemory: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "proc_memory"),
+			"Memory consumed by a process",
+			[]string{"name", "pid"},
+			nil,
+		),
+		procStatus: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "proc_status"),
+			"Running status for a process.",
+			[]string{"name", "pid", "codeRevision", "lifeStatus", "enabled"},
+			nil,
+		),
+	}
+}
+
+// Collect fetches the statistics from the configured memcached server, and
+// delivers them as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	info, err := e.status()
+	if err != nil {
+		ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 0)
+		log.Errorf("failed to collect status from passenger: %s", err)
+		return
+	}
+	ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 1)
+
+	// TODO: Doesn't seem to be a queryable uptime on the main process.
+	// Talk with @grobie about how to handle this.
+
+	ch <- prometheus.MustNewConstMetric(e.version, prometheus.GaugeValue, 1, info.PassengerVersion)
+
+	// This value should be 0. Not sure if I should use a gauge or a
+	// counter, as we shouldn't be counting anything.
+	ch <- prometheus.MustNewConstMetric(e.toplevelQueue, prometheus.CounterValue, parseFloat(info.TopLevelRequestsInQueue))
+	ch <- prometheus.MustNewConstMetric(e.maxProcessCount, prometheus.GaugeValue, parseFloat(info.MaxProcessCount))
+	ch <- prometheus.MustNewConstMetric(e.currentProcessCount, prometheus.GaugeValue, parseFloat(info.CurrentProcessCount))
+	ch <- prometheus.MustNewConstMetric(e.appCount, prometheus.GaugeValue, parseFloat(info.AppCount))
+
+	for _, sg := range info.SuperGroups {
+		ch <- prometheus.MustNewConstMetric(e.appQueue, prometheus.GaugeValue, parseFloat(sg.RequestsInQueue), sg.Name)
+		ch <- prometheus.MustNewConstMetric(e.appProcsSpawning, prometheus.GaugeValue, parseFloat(sg.Group.ProcessesSpawning), sg.Name)
+
+		for _, proc := range sg.Group.Processes {
+			ch <- prometheus.MustNewConstMetric(e.procMemory, prometheus.GaugeValue, parseFloat(proc.RealMemory), sg.Name, proc.PID)
+			ch <- prometheus.MustNewConstMetric(e.requestsProcessed, prometheus.CounterValue, parseFloat(proc.RequestsProcessed), sg.Name, proc.PID)
+
+			if uptime, err := parsePassengerInterval(proc.Uptime); err == nil {
+				ch <- prometheus.MustNewConstMetric(e.procUptime, prometheus.CounterValue, float64(uptime), sg.Name, proc.PID)
+			}
+
+			// Is this one really necessary?
+			ch <- prometheus.MustNewConstMetric(
+				e.procStatus, prometheus.CounterValue, 1,
+				sg.Name, proc.PID, proc.CodeRevision, proc.LifeStatus, proc.Enabled,
+			)
+		}
+	}
+
+}
+
+func (e *Exporter) status() (*Info, error) {
+	var (
+		out bytes.Buffer
+		cmd = exec.Command(e.cmd, e.args...)
```

Comment:
Yes, any error should result in passenger_up=0. (up will be set by
Prometheus).

On Tue, May 10, 2016 at 5:02 PM, stuart nelson <notifications@github.com>
wrote:

> In main.go
> <https://github.com/stuartnelson3/passenger_exporter/pull/1#discussion_r62751755>
> :
>
> > +			}
> > +
> > +			// Is this one really necessary?
> > +			ch <- prometheus.MustNewConstMetric(
> > +				e.procStatus, prometheus.CounterValue, 1,
> > +				sg.Name, proc.PID, proc.CodeRevision, proc.LifeStatus, proc.Enabled,
> > +			)
> > +		}
> > +	}
> > +
> > +}
> > +
> > +func (e *Exporter) status() (*Info, error) {
> > +	var (
> > +		out bytes.Buffer
> > +		cmd = exec.Command(e.cmd, e.args...)
>
> I think we should definitely set up a timeout, but in the event of a
> timeout we should probably do something. If passenger-status can't
> respond within, oh, 500ms, I would say something is very wrong with it.
> Return a timeout error, and then let Collect() set up=0?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/stuartnelson3/passenger_exporter/pull/1/files/e22093294ce73447f7eadfb1828ec2f4893a1c68#r62751755>
>